src/Build/BackEnd/TaskExecutionHost/TaskExecutionHost.cs

```diff
@@ -1395,9 +1395,24 @@ private void GatherTaskItemOutputs(bool outputTargetIsItem, string outputTargetN
                                     // Setting an item spec expects the escaped value, as does setting metadata.
                                     newItem = new ProjectItemInstance(_projectInstance, outputTargetName, EscapingUtilities.Escape(output.ItemSpec), parameterLocationEscaped);
 
-                                    newItem.SetMetadataOnTaskOutput(output.CloneCustomMetadata()
-                                        .Cast<DictionaryEntry>()
-                                        .Select(x => new KeyValuePair<string, string>((string)x.Key, EscapingUtilities.Escape((string)x.Value))));
+                                    var items = new List<KeyValuePair<string, string>>();
+                                    foreach (object entry in output.CloneCustomMetadata())
+                                    {
+                                        if (entry is DictionaryEntry de)
+                                        {
+                                            items.Add(new KeyValuePair<string, string>((string)de.Key, EscapingUtilities.Escape((string)de.Value)));
+                                        }
+                                        else if (entry is KeyValuePair<string, string> kvp)
+                                        {
+                                            items.Add(new KeyValuePair<string, string>(kvp.Key, EscapingUtilities.Escape(kvp.Value)));
+                                        }
+                                        else
+                                        {
+                                            throw new InternalErrorException("Metadata item was neither DictionaryEntry nor KeyValuePair<string, string>");
+                                        }
+                                    }
```

Comment:
You don't have to do all this--in the `foreach`, you're guaranteed to get the `DictionaryEntry`.

Reply:
Outdated